mesonbuild/interpreter/interpreter.py

```diff
@@ -1730,39 +1732,64 @@ def func_disabler(self, node, args, kwargs):
     @FeatureNewKwargs('executable', '0.56.0', ['win_subsystem'])
     @FeatureDeprecatedKwargs('executable', '0.56.0', ['gui_app'], extra_message="Use 'win_subsystem' instead.")
     @permittedKwargs(build.known_exe_kwargs)
-    def func_executable(self, node, args, kwargs):
+    @typed_pos_args('executable', str, varargs=(str, mesonlib.File, build.CustomTarget, build.CustomTargetIndex, build.GeneratedList, build.StructuredSources))
+    def func_executable(self, node: mparser.BaseNode,
```

Comment:
From a quick glance, I think we only ever validated this list before in `source_strings_to_files()`. 

In theory that permits build.BuildTarget and build.ExtractedObjects too, how did we used to handle this?

Reply:
To answer my own question -- both of them work, but produce:

```
WARNING: Build target bar has no sources. This was never supposed to be allowed but did because of a bug, support will be removed in a future release of Meson
```

It even works ***without warnings*** if you specify a proper source too.

Reply:
Welll... looking teep into the depths of `build.py`, I found:
```python
    def process_sourcelist(self, sources: T.List['SourceOutputs']) -> None:
        """Split sources into generated and static sources.

        Sources can be:
        1. Pre-existing source files in the source tree (static)
        2. Pre-existing sources generated by configure_file in the build tree.
           (static as they are only regenerated if meson itself is regenerated)
        3. Sources files generated by another target or a Generator (generated)
        """
        added_sources: T.Set[File] = set() # If the same source is defined multiple times, use it only once.
        for s in sources:
            if isinstance(s, File):
                if s not in added_sources:
                    self.sources.append(s)
                    added_sources.add(s)
            elif isinstance(s, (CustomTarget, CustomTargetIndex, GeneratedList)):
                self.generated.append(s)
```

Sooooo... they're totally allowed, and then silently dropped on the floor.

Reply:
So I guess we need to allow that then deprecate it